common/src/main/java/org/conscrypt/OpenSSLX25519PrivateKey.java

```diff
@@ -1,51 +1,60 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.conscrypt;
 
 import java.security.PrivateKey;
+import java.security.spec.EncodedKeySpec;
 import java.security.spec.InvalidKeySpecException;
-import java.security.spec.PKCS8EncodedKeySpec;
 import java.util.Arrays;
 
 public class OpenSSLX25519PrivateKey implements OpenSSLX25519Key, PrivateKey {
     private static final long serialVersionUID = -3136201500221850916L;
     private static final byte[] PKCS8_PREAMBLE = new byte[]{
-            0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x6e, 0x04, 0x22, 0x04, 0x20,
-    };
-
-    private static final byte[] PKCS8_PREAMBLE_WITH_NULL = new byte[] {
-            0x30, 0x30, 0x02, 0x01, 0x00, 0x30, 0x07, 0x06, 0x03, 0x2B, 0x65, 0x6E, 0x05, 0x00, 0x04, 0x22, 0x04, 0x20,
+            0x30, 0x2e,                            // Sequence: 46 bytes
+                0x02, 0x01, 0x00,                  // Integer: 0 (version)
+                0x30, 0x05,                        // Sequence: 5 bytes
+                    0x06, 0x03, 0x2b, 0x65, 0x6e,  // OID: 1.3.101.110 (X25519)
+                0x04, 0x22, 0x04, 0x20,            // Octet string: 32 bytes
+            // Key bytes follow directly
     };
 
     private byte[] uCoordinate;
 
-    public OpenSSLX25519PrivateKey(PKCS8EncodedKeySpec keySpec) throws InvalidKeySpecException {
+    public OpenSSLX25519PrivateKey(EncodedKeySpec keySpec) throws InvalidKeySpecException {
         byte[] encoded = keySpec.getEncoded();
-        if (encoded == null || !"PKCS#8".equals(keySpec.getFormat())) {
-            throw new InvalidKeySpecException("Key must be encoded in PKCS#8 format");
+        if ("PKCS#8".equals(keySpec.getFormat())) {
+            if (!ArrayUtils.startsWith(encoded, PKCS8_PREAMBLE)) {
```

Comment:
Existing issue but PrivateKeyInfo does have an optional list of attributes in it, so this code will reject some valid PKCS#8 X25519 keys. (BoringSSL's PKCS#8 parser will handle this, if you want to call into that. It's `EVP_parse_private_key`.)

SubjectPublicKeyInfo only has one valid encoding, so the prefix trick is perfectly OK on the public key side.

Reply:
Ah thanks! I think I misremembered the offline conversation  we had about this and assumed that because the optional attributes aren't recommended then we could reject keys using them (which now I write it down sounds silly, so I'm sure that's not what you said).

`EVP_parse_private_key` is probably the best way forward.

Reply:
Latest commit switches to BoringSSL's PKCS#8 parser, would appreciate it if you could give the JNI a once over.

There's a bit of a rabbit hole with lots of places throwing an internal ParseException where we could probably do better, but rather than address that in this CL I have added one more usage of it to review later :/

Reply:
Oh, I mean, that's also an existing issue, so I assumed you'd do it separately. Happy to review it here though. :-)

Reply:
Thanks!